### PR TITLE
 [app sec] fix bug when debug logs are enabled

### PR DESCRIPTION
**Pull request by robertpi**
Fix a bug where app sec failed in debug log mode because of formatting a value

Also:

- Fix an issue with build_in_docker.sh
- Move some logs over to structured logging

Fixes #

Changes proposed in this pull request:

@DataDog/apm-dotnet

### CODE REVIEW

File: tracer/build/_build/docker/debian.dockerfile [hunk 1]
```diff
@@ -20,7 +20,8 @@ RUN apt-get update \
         ruby \
         ruby-dev \
         rubygems \
-    && gem install --no-document fpm
+    && gem install --no-document fpm \
+    && dotnet dev-certs https
```
Comment (robertpi) [inline tracer/build/_build/docker/debian.dockerfile:24]: Not sure why this necessary, when I've never had to do it before, but got an error message when starting a web app saying it was necessary.

File: tracer/build_in_docker.sh [hunk 2]
```diff
@@ -1,8 +1,6 @@
 #!/usr/bin/env bash
 set -euox pipefail
 
-# in case we are being run from outside this directory
```
Comment (robertpi) [inline tracer/build_in_docker.sh:4]: The only way I found to get build_in_docker.sh to works was to remove this line and run it from the source tree root.

Comment (andrewlock) [reply tracer/build_in_docker.sh:4]: Please don't remove this, the correct fix is shown in https://github.com/DataDog/dd-trace-dotnet/commit/943dc89623c825dfd6dc0dee1609c486cdba9143

This was broken by the tracer directory move

File: tracer/build/_build/docker/debian.dockerfile (hunk 1 above)
Comment (andrewlock) [reply tracer/build/_build/docker/debian.dockerfile:24]: I don't think this should be necessary, given we have passing builds in CI? 🤷

File: tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs [hunk 3]
```diff
@@ -91,7 +91,7 @@ o switch
                     IList<JToken> objs => EncodeList(objs.Select(x => (object)x), argCache, remainingDepth),
                     IList<string> objs => EncodeList(objs.Select(x => (object)x), argCache, remainingDepth),
                     IList<object> objs => EncodeList(objs, argCache, remainingDepth),
-                    _ => throw new Exception($"Couldn't encode: {o}, type: {o.GetType()}")
+                    _ => throw new Exception($"Couldn't encod type: {o?.GetType()}")
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs:94]:
```suggestion
                    _ => throw new Exception($"Couldn't encode type: {o?.GetType()}")
```

**Review by andrewlock — COMMENTED** [review]
a few nit comments, I'm not sure which of the logging changes is the bug fix?

File: tracer/build/_build/docker/debian.dockerfile (hunk 1 above)
Comment (robertpi) [reply tracer/build/_build/docker/debian.dockerfile:24]: I agree, it's weird, but after a `docker system prune --all`, running:

```
./tracer/build_in_docker.sh BuildTracerHome && \
   ./tracer/build_in_docker.sh RunDotNetSample  -framework net5.0 -SampleName Samples.AspNetCore5

```
fails to start the web server with the attached log:

[RunDotNetSample-failure.txt](https://github.com/DataDog/dd-trace-dotnet/files/7202324/RunDotNetSample-failure.txt)

I'm pretty sure this did used to work without need to call, `dotnet dev-certs https`, but I've no idea what changed.

File: tracer/build_in_docker.sh (hunk 2 above)
Comment (robertpi) [reply tracer/build_in_docker.sh]: reverted in e39dbcff05c2af9cd990914568938d6793651134

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/630bbc93132e062d36cfef25f1e498f787935af5) will **not change** line coverage
:heavy_check_mark: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/630bbc93132e062d36cfef25f1e498f787935af5) will **not change** branch coverage
:no_entry: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/630bbc93132e062d36cfef25f1e498f787935af5) will will **increase** complexity by `12`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/630bbc93132e062d36cfef25f1e498f787935af5) | #1796       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9329` / `14796` | `9329` / `14804` |          |
| Lines %   | `63%`      | `63%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4137` / `6452` | `4142` / `6464` |          |
| Branches %| `64%`      | `64%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7596`      | `7608`        |  `12`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75535&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75535/artifacts?artifactName=Code%20Coverage%20Report_75535&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1796](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75537&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75537/artifacts?artifactName=Code%20Coverage%20Report_75537&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/630bbc93132e062d36cfef25f1e498f787935af5) | #1796       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `63%`      | `63%`       |  `0%` :heavy_check_mark:  |
| Branches %| `64%`      | `64%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7596`      | `7608`       |  `12` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Logging.LogRateLimiter](https://github.com/DataDog/dd-trace-dotnet/tree/6d7d6ec14810b99e7aad7a80eae335e3e2e0dcfe/tracer/src/Datadog.Trace/Logging/LogRateLimiter.cs) | `0%` :heavy_check_mark: | `17%` :heavy_check_mark:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75535&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75535/artifacts?artifactName=Code%20Coverage%20Report_75535&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1796](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75537&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75537/artifacts?artifactName=Code%20Coverage%20Report_75537&api-version=6.0&%24format=zip)

File: tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs [hunk 4]
```diff
@@ -198,10 +198,14 @@ o switch
...
-                    IEnumerable<KeyValuePair<string, object>> objDict => FormatDictionary(objDict, sb),
+                    IEnumerable<KeyValuePair<string, JToken>> objDict => FormatDictionary(objDict.Select(x => new KeyValuePair<string, object>(x.Key, x.Value)), sb),
                     IEnumerable<KeyValuePair<string, string>> objDict => FormatDictionary(objDict.Select(x => new KeyValuePair<string, object>(x.Key, x.Value)), sb),
-                    _ => throw new Exception($"Couldn't encode: {o}, type: {o.GetType()}")
+                    IEnumerable<KeyValuePair<string, List<string>>> objDict => FormatDictionary(objDict.Select(x => new KeyValuePair<string, object>(x.Key, x.Value)), sb),
+                    IEnumerable<KeyValuePair<string, object>> objDict => FormatDictionary(objDict, sb),
+                    IList<JToken> objs => FormatList(objs.Select(x => (object)x), sb),
+                    IList<string> objs => FormatList(objs.Select(x => (object)x), sb),
+                    IList<object> objs => FormatList(objs, sb),
+                    _ => sb.Append($"Error: couldn't format type: {o?.GetType()}")
```
Comment (robertpi) [inline tracer/src/Datadog.Trace/AppSec/Waf/Encoder.cs:208]: These changes fix the initial bug, the formatter needs to be able to handle the same cases as the encoder. Also, decided it's not helpful to throw and error here.

File: tracer/src/Datadog.Trace.ClrProfiler.Native/interop.cpp [hunk 5]
```diff
@@ -37,14 +36,7 @@ EXTERN_C void *dddlopen (const char *__file, int __mode)
 
 EXTERN_C char *dddlerror (void)
 {
-    auto errorPtr = dlerror();
```
Comment (robertpi) [inline tracer/src/Datadog.Trace.ClrProfiler.Native/interop.cpp:40]: @andrewlock this corrects the excessive error logs we were discussing.

**Review by andrewlock — APPROVED** [review]

**Review by zacharycmontoya — APPROVED** [review]

**Comment by andrewlock** [comment]
## Code Coverage Report :bar_chart:

:heavy_check_mark: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c5b1e91c8b5997ca4187199d0bc2ee51b12ab2fc) will **not change** line coverage
:heavy_check_mark: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c5b1e91c8b5997ca4187199d0bc2ee51b12ab2fc) will **not change** branch coverage
:no_entry: Merging #1796 into [master](https://github.com/DataDog/dd-trace-dotnet/tree/c5b1e91c8b5997ca4187199d0bc2ee51b12ab2fc) will will **increase** complexity by `12`


|           | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c5b1e91c8b5997ca4187199d0bc2ee51b12ab2fc) | #1796       | Change   | 
|:----------|:-----------:|:-----------:|:--------:|
| Lines     | `9429` / `15031` | `9428` / `15040` |          |
| Lines %   | `63%`      | `63%`      |  `0%` :heavy_check_mark:  |
| Branches  | `4157` / `6510` | `4158` / `6522` |          |
| Branches %| `64%`      | `64%`      |  `0%` :heavy_check_mark:  |
| Complexity|   `7672`      | `7684`        |  `12`  :no_entry:    |

View the full report for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75881&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75881/artifacts?artifactName=Code%20Coverage%20Report_75881&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1796](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75910&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75910/artifacts?artifactName=Code%20Coverage%20Report_75910&api-version=6.0&%24format=zip)

### Datadog.Trace Breakdown :heavy_check_mark:

|        | [master](https://github.com/DataDog/dd-trace-dotnet/tree/c5b1e91c8b5997ca4187199d0bc2ee51b12ab2fc) | #1796       | Change   | 
|:-------|:-----------:|:-----------:|:--------:|
| Lines %| `63%`      | `63%`       |  `0%` :heavy_check_mark:  |
| Branches %| `64%`      | `64%`       |  `0%` :heavy_check_mark:  |
| Complexity| `7672`      | `7684`       |  `12` :no_entry:  |

The following classes have significant coverage changes.

| File    | Line coverage change | Branch coverage change | Complexity change |
|:--------|:--------------------:|:----------:|:--------:|
| [Datadog.Trace.Agent.Transports.HttpStreamResponse](https://github.com/DataDog/dd-trace-dotnet/tree/099e49bb8876adf59458ef9599872277d8aa4318/tracer/src/Datadog.Trace/Agent/Transports/HttpStreamResponse.cs) | `-10%` :no_entry: | `0%` :heavy_check_mark:   | `0` :heavy_check_mark: |
| [Datadog.Trace.HttpOverStreams.HttpHeaders](https://github.com/DataDog/dd-trace-dotnet/tree/099e49bb8876adf59458ef9599872277d8aa4318/tracer/src/Datadog.Trace/HttpOverStreams/HttpHeaders.cs) | `-4%` :warning: | `-12%` :no_entry:   | `0` :heavy_check_mark: |


View the full reports for further details:

* [HTML report for master](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75881&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75881/artifacts?artifactName=Code%20Coverage%20Report_75881&api-version=6.0&%24format=zip)  
* [HTML report for this PR #1796](https://dev.azure.com/datadoghq/dd-trace-dotnet/_build/results?buildId=75910&view=codecoverage-tab) | [Source file](https://dev.azure.com/datadoghq/a51c4863-3eb4-4c5d-878a-58b41a049e4e/_apis/build/builds/75910/artifacts?artifactName=Code%20Coverage%20Report_75910&api-version=6.0&%24format=zip)

**Pull request merged**